errors/error.go

```diff
@@ -0,0 +1,373 @@
+// There must be at least one argument or E panics.
+// The type of each argument determines its meaning. If more than one argument
+// of a given type is presented, only the last one is recorded.
+//
+// The types are:
+// 	errors.Kind
+//		The kind of error (eg.: HCLSyntax, TerramateSchema, etc).
+//	hcl.Range
+//		The file range where the error originated.
+//  errors.Stack
+//		The stack which the error originated.
+//	error
+//		The underlying error that triggered this one.
+//	hcl.Diagnostics
+//		The underlying hcl error that triggered this one.
+//		If this error's FileRange is not set, the diagnostic subject range is
+//		pulled.
+//		If this error's Description is not set, the diagnostic detail field is
+//		pulled.
+//	string
+//		The error description. It supports formatting using the Go's fmt verbs
+//      as long as the arguments are not of the defined types.
+//
+// If the error is printed, only those items that have been
+// set to non-zero values will appear in the result. For the `hcl.Range` type,
+// the `range.Empty()` method is used.
+//
+// Error promotions:
+//
+// If Kind is not specified or Any, we set it to the Kind of
+// the underlying error.
+// If FileRange is not specified, we set it to the FileRange of the underlying
+// error (if set).
+//
+// Minimization:
+//
+// In order to avoid duplicated messages, we erase the fields present in the
+// underlying error if already set with same value in this error.
+func E(args ...interface{}) error {
+	if len(args) == 0 {
+		panic("called with no args")
+	}
+
+	var (
+		diags  hcl.Diagnostics
+		format *string
+	)
+
+	fmtargs := []interface{}{}
+
+	e := &Error{}
+	for _, arg := range args {
+		switch arg := arg.(type) {
+		case Kind:
+			e.Kind = arg
+		case hcl.Range:
+			e.FileRange = arg
+		case hcl.Diagnostics:
+			diags = arg
+			if diags.HasErrors() {
+				diag := diags[0]
+				if diag.Subject != nil {
+					e.FileRange = *diag.Subject
+				}
+			}
+		case StackMeta:
+			e.Stack = arg
+		case error:
+			e.Err = arg
+		case string:
+			val := arg
+			if format == nil {
+				format = &val
+			} else {
+				fmtargs = append(fmtargs, val)
+			}
+		default:
+			fmtargs = append(fmtargs, arg)
+		}
+	}
+
+	if format != nil {
+		e.Description = fmt.Sprintf(*format, fmtargs...)
+	} else if len(fmtargs) > 0 {
+		panic(errors.New("errors.E called with arbitraty types and no format"))
+	}
+
+	if diags.HasErrors() {
+		if e.Description == "" {
+			e.Description = diags[0].Detail
+		} else if e.Err == nil {
+			e.Err = diags
+		}
+	}
+
+	if e.isEmpty() {
+		panic(errors.New("empty error"))
+	}
+	prev, ok := e.Err.(*Error)
+	if !ok {
+		return e
+	}
+	if prev.Kind == e.Kind {
+		prev.Kind = ""
+	}
+	if e.Kind == "" {
+		e.Kind = prev.Kind
+		prev.Kind = ""
+	}
+
+	emptyRange := hcl.Range{}
+	if e.FileRange == emptyRange {
+		e.FileRange = prev.FileRange
+	}
+	if prev.FileRange == e.FileRange {
+		prev.FileRange = hcl.Range{}
+	}
+	if prev.Stack == e.Stack {
+		prev.Stack = nil
+	}
```

Comment:
Also, feel free to disregard this because maybe Im being pedantic, but comparing using interfaces here seems risky if there is any other field on the underlying object that implements the interface that may differ and we are not interested in. For our purposes we are interested on then having the same name/path/description and that is it. Also, even though it is a stretch, if there is any type implementing this interface that is not comparable this would panic.

When doing the error matching we are already comparing by calling the interface methods: https://github.com/mineiros-io/terramate/pull/291/files#diff-d128886dfda28afbfb957c032a184b95f625ca12ff25c8656247783f3e9bf112R364

So maybe we could do it here too.

Reply:
this comparison is there since errors.Stack was a string. It needs to be fixed. Thanks for catching it.